### PR TITLE
SDK maintenance

### PR DESCRIPTION
**Pull request by dennisklein**
Working on outstanding issues and quality of new SDK components:

* [x] Adapt to recent API changes in DDS `2.5.24+` (solved by #200)
* [x] Refactor `sdk::Topology::ChangeState` implementation to use asio primitives and normalized types
* [x] Improve doxygen commentary
* ~Introduce `fairmq-top` which polls topology state via `sdk::Topology` and prints a ncurses-based table view similar to `htop`~
   * ~This will be a useful utility but also helps further development of the SDK because it will benchmark the SDK APIs in performance, usability, and correctness.~ *split into a separate PR #203*
* Comply with asio's [AsynchronousOperation concept](https://www.boost.org/doc/libs/develop/doc/html/boost_asio/reference/asynchronous_operations.html) in
  * ~`fairmq-top`~ *split to a separate PR #203*
  * [x] `sdk::Topology::Async*`
  * ~`sdk::DDS*::Async*` (Implement and comply)~ *not needed for now*
* ~#191~ *postponed, do not know a good solution yet*
* ~#192~ *postponed to a later PR*

### CODE REVIEW

File: fairmq/sdk/top/main.cxx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        // try {
+            // future.get();
+            // rc = EXIT_SUCCESS;
+        // } catch (const system::error_code& e) {
+            // rc = EXIT_FAILURE;
+        // }
+
+        // context.stop();
+        // ctxThread.join();
+        //
```
Comment (dennisklein) [inline fairmq/sdk/top/main.cxx]: @mattkretz Trying to comply with asio's [AsynchronousOperation](https://www.boost.org/doc/libs/1_70_0/doc/html/boost_asio/reference/asynchronous_operations.html) concept. Calling with a completion token in the form of a lambda, [my code](https://github.com/FairRootGroup/FairMQ/pull/195/files#diff-3dfc84dcdb355ed30b83b782ee6d39e0R104) compiles and works as expected. Using the `boost::asio::use_future` completion token instead like this (in https://github.com/FairRootGroup/FairMQ/pull/195/files#diff-3dfc84dcdb355ed30b83b782ee6d39e0R123):

```cpp
        std::thread ctxThread([&]{ context.run(); });

        auto future = app.AsyncRun(asio::use_future);

        try {
            future.get();
            rc = EXIT_SUCCESS;
        } catch (const system::error_code& e) {
            rc = EXIT_FAILURE;
        }

        context.stop();
        ctxThread.join();
```

leads to the following compiler error:

```
[1/3] Building CXX object fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
FAILED: fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
/usr/bin/c++  -DBOOST_ERROR_CODE_HEADER_ONLY -DFAIRLOGGER_USE_BOOST_PRETTY_FUNCTION -I../ -I. -isystem /home/dklein/boost_1_70_0_install/include -isystem /home/dklein/projects/FairLogger/build/install/include -fdiagnostics-color=always -O2 -g -Wshadow -Wall -Wextra -DNDEBUG -fPIE   -pthread -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -std=c++14 -MD -MT fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -MF fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o.d -o fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -c ../fairmq/sdk/top/main.cxx
In file included from ../fairmq/sdk/top/main.cxx:15:
../fairmq/sdk/top/Top.h: In instantiation of ‘fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)> [with auto:1 = boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >]’:
/home/dklein/boost_1_70_0_install/include/boost/asio/async_result.hpp:273:47:   required from ‘typename std::enable_if<(! boost::asio::detail::async_result_has_initiate_memfn<CompletionToken, Signature>::value), typename boost::asio::async_result<typename std::decay<_Tp>::type, Signature>::return_type>::type boost::asio::async_initiate(Initiation&&, CompletionToken&, Args&& ...) [with CompletionToken = const boost::asio::use_future_t<>&; Signature = void(boost::system::error_code); Initiation = fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)>; Args = {}; typename std::enable_if<(! boost::asio::detail::async_result_has_initiate_memfn<CompletionToken, Signature>::value), typename boost::asio::async_result<typename std::decay<_Tp>::type, Signature>::return_type>::type = std::future<void>]’
../fairmq/sdk/top/Top.h:86:79:   required from ‘auto fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]’
../fairmq/sdk/top/main.cxx:123:52:   required from here
../fairmq/sdk/top/Top.h:93:26: error: no match for call to ‘(const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >) (boost::system::error_code)’
   93 |                         h(system::errc::make_error_code(system::errc::success));
      |                         ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/dklein/boost_1_70_0_install/include/boost/asio/use_future.hpp:157,
                 from ../fairmq/sdk/top/main.cxx:13:
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note: candidate: ‘void boost::asio::detail::promise_handler_ec_0::operator()(const boost::system::error_code&)’ <near match>
  303 |   void operator()(const boost::system::error_code& ec)
      |        ^~~~~~~~
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note:   passing ‘const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >*’ as ‘this’ argument discards qualifiers
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:228:8: error: ‘void boost::asio::detail::promise_executor<T>::post(F&&, const A&) const [with F = fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)> [with auto:1 = boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >]::<lambda()>; A = std::allocator<void>; T = void]’, declared using local type ‘fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)> [with auto:1 = boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >]::<lambda()>’, is used but never defined [-fpermissive]
  228 |   void post(BOOST_ASIO_MOVE_ARG(F) f, const A& a) const
      |        ^~~~
ninja: build stopped: subcommand failed.
```

The mentioned `note: candidate: ‘void boost::asio::detail::promise_handler_ec_0::operator()(const boost::system::error_code&)’ <near match>` is what I expected to be available, but something not obvious to me is preventing my implementation from choosing it. You can find some of the relevant templates here:
* `../fairmq/sdk/top/Top.h:93` [`fair::mq::sdk::Top::AsyncRun`](https://github.com/FairRootGroup/FairMQ/pull/195/files#diff-556ca1f50fc2fcbd48924322db6e606cR93)
* [`boost::asio::detail::promise_handler`](https://github.com/boostorg/asio/blob/boost-1.70.0/include/boost/asio/impl/use_future.hpp#L589) and its base classes are defined in the same file above.

Would be awesome if you could have a quick look, thx!

Comment (dennisklein) [reply fairmq/sdk/top/main.cxx]: Implementing `sdk::Top::AsyncRun` without the `boost::asio::async_initiate` helper like this

```cpp
    template<typename CompletionToken>
    auto AsyncRun(CompletionToken&& token)
    {
        using namespace boost;

        asio::async_completion<CompletionToken, void(system::error_code)> completion(token);

        auto ex(asio::get_associated_executor(completion.completion_handler, fExecutor));
        auto alloc(asio::get_associated_allocator(completion.completion_handler));

        ex.post(
            [h = std::move(completion.completion_handler)]() {
                h(system::errc::make_error_code(system::errc::success));
            },
            alloc);

        return completion.result.get();

        // return asio::async_initiate<CompletionToken, void(system::error_code)>(
            // [&](auto handler) {
                // auto ex(asio::get_associated_executor(handler, fExecutor));
                // auto alloc(asio::get_associated_allocator(handler));
//
                // ex.post(
                    // [h = std::move(handler)]() {
                        // h(system::errc::make_error_code(system::errc::success));
                    // },
                    // alloc);
            // },
            // token);
    }
```

condenses the compiler error message a bit:

```
[1/3] Building CXX object fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
FAILED: fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
/usr/bin/c++  -DBOOST_ERROR_CODE_HEADER_ONLY -DFAIRLOGGER_USE_BOOST_PRETTY_FUNCTION -I../ -I. -isystem /home/dklein/boost_1_70_0_install/include -isystem /home/dklein/projects/FairLogger/build/install/include -fdiagnostics-color=always -O2 -g -Wshadow -Wall -Wextra -DNDEBUG -fPIE   -pthread -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -std=c++14 -MD -MT fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -MF fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o.d -o fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -c ../fairmq/sdk/top/main.cxx
In file included from ../fairmq/sdk/top/main.cxx:15:
../fairmq/sdk/top/Top.h: In instantiation of ‘auto fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]’:
../fairmq/sdk/top/main.cxx:123:52:   required from here
../fairmq/sdk/top/Top.h:93:18: error: no match for call to ‘(const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >) (boost::system::error_code)’
   93 |                 h(system::errc::make_error_code(system::errc::success));
      |                 ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/dklein/boost_1_70_0_install/include/boost/asio/use_future.hpp:157,
                 from ../fairmq/sdk/top/main.cxx:13:
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note: candidate: ‘void boost::asio::detail::promise_handler_ec_0::operator()(const boost::system::error_code&)’ <near match>
  303 |   void operator()(const boost::system::error_code& ec)
      |        ^~~~~~~~
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note:   passing ‘const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >*’ as ‘this’ argument discards qualifiers
ninja: build stopped: subcommand failed.
```

Comment (mattkretz) [reply fairmq/sdk/top/main.cxx]: your `operator()` is missing a const qualification. i.e. `operator()(...) const`

Comment (mattkretz) [reply fairmq/sdk/top/main.cxx]: ah well, it's asio's `operator()` that's non-const. So your problem is that `promise_handler` must be non-const.

Comment (dennisklein) [reply fairmq/sdk/top/main.cxx]: thx!

**Comment by dennisklein** [comment]
@mattkretz Would be awesome if you could review the `fairmq/sdk/AsioAsyncOp.h` file in this PR if you can spare the time. Thx! (Some tests are in `test/sdk/_async_op.cxx`)

File: test/protocols/_push_pull_multipart.cxx [hunk 2]
```diff
@@ -33,9 +33,9 @@ auto RunSingleThreadedMultipart(string transport, string address) -> void {
     config.SetProperty<string>("session", std::to_string(session));
     auto factory = FairMQTransportFactory::CreateTransportFactory(transport, fair::mq::tools::Uuid(), &config);
     FairMQTransportFactory* factoryptr = factory.get();
-    auto push = FairMQChannel{"Push", "push", factory};
+    FairMQChannel push("Push", "push", factory);
```
Comment (rbx) [inline test/protocols/_push_pull_multipart.cxx]: :+1:

File: fairmq/sdk/Topology.h [hunk 3]
```diff
@@ -81,82 +86,325 @@ inline bool StateEqualsTo(const TopologyState& topologyState, DeviceState state)
...
+                        fChangeStateOpTimer.expires_after(timeout);
+                        fChangeStateOpTimer.async_wait([&](std::error_code ec) {
+                            if (!ec) {
+                                std::lock_guard<std::mutex> lk2(fMtx);
+                                fChangeStateOp.Timeout(fState);
+                            }
+                        });
+                    }
+                } else {
+                    throw std::system_error(MakeErrorCode(ErrorCode::OperationInProgress),
```
Comment (dennisklein) [inline fairmq/sdk/Topology.h]: @rbx Just realized this should no longer be an exception but an immediate completion with the same error code. Missed that one before.

Comment (rbx) [reply fairmq/sdk/Topology.h]: Do you want to change this in this PR?

Comment (dennisklein) [reply fairmq/sdk/Topology.h]: yes

**Comment by dennisklein** [comment]
@rbx This is ready now, but I want another round of testing on different machines and on mac before I merge it. Also find the new `sdk_tester` branch which delivers a minimal example on how to consume the sdk (sync for now - added it mainly for testing the cmake exports).

File: fairmq/sdk/Topology.h (hunk 3 above)
Comment (dennisklein) [reply fairmq/sdk/Topology.h]: fixed now

**Comment by dennisklein** [comment]
Merging, tested both on mac and linux extensively.

**Pull request merged**